### PR TITLE
SONARJAVA-4052 Provide quick fix availability to SQ

### PR DESCRIPTION
**Pull request by mpaladin**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/reporting/InternalJavaIssueBuilder.java [hunk 1]
```diff
@@ -227,10 +227,12 @@ public void report() {
       }
     }
 
-
     if (!quickFixes.isEmpty() && sonarComponents.isQuickFixCompatible()) {
       addQuickFixes(ruleKey.get(), (NewSonarLintIssue) newIssue);
     }
+    if (!quickFixes.isEmpty() && sonarComponents.hasMethodSetQuickFixAvailable()) {
```
[the remark this answers is not in the document]
Comment (mpaladin) [reply java-frontend/src/main/java/org/sonar/java/reporting/InternalJavaIssueBuilder.java:233]: @quentin-jaquier-sonarsource ok, thank you for the note.

File: java-frontend/src/main/java/org/sonar/java/reporting/InternalJavaIssueBuilder.java [hunk 2]
```diff
@@ -227,32 +232,40 @@ public void report() {
       }
     }
 
-
-    if (!quickFixes.isEmpty() && sonarComponents.isQuickFixCompatible()) {
-      addQuickFixes(ruleKey.get(), (NewSonarLintIssue) newIssue);
+    final List<JavaQuickFix> flatQuickFixes = quickFixes.stream()
+      .flatMap(s -> s.get().stream())
+      .collect(Collectors.toList());
```
Comment (quentin-jaquier-sonarsource) [inline java-frontend/src/main/java/org/sonar/java/reporting/InternalJavaIssueBuilder.java:237]: As discussed, we are always computing the quick fixes here, even when we have nothing to do with them.
I agree that we will eventually always compute them for all products and delaying the computation is just a temporary minor performance gain. Still, we already have the logic in place, we can benefit from it by refactoring the logic here.
Even if it is less readable, I think it is worth it.

File: java-frontend/src/main/java/org/sonar/java/reporting/InternalJavaIssueBuilder.java [hunk 3]
```diff
@@ -227,32 +232,40 @@ public void report() {
...
+    if (!flatQuickFixes.isEmpty()) {
+      if (isQuickFixCompatible) {
+        addQuickFixes(inputFile, ruleKey.get(), flatQuickFixes, (NewSonarLintIssue) newIssue);
+      } else if (methodSetQuickFixAvailable != null) {
+        try {
+          methodSetQuickFixAvailable.invoke(newIssue, true);
+        } catch (ReflectiveOperationException e) {
+          throw new RuntimeException(e);
```
Comment (quentin-jaquier-sonarsource) [inline java-frontend/src/main/java/org/sonar/java/reporting/InternalJavaIssueBuilder.java:245]: Throwing this exception here will stop the analysis of the whole file, even if it is unlikely, it seems a bit overkill.

**Review by quentin-jaquier-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**